File: SKIBin_old/app/src/main/java/com/h2Invent/skibin/FingerprintActivity.java
package com.h2Invent.skibin;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.biometric.BiometricPrompt;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Random;
import java.util.concurrent.Executor;

public class FingerprintActivity extends AppCompatActivity {
    private static String SECURITY_SAVE = "SECURITY_SAVE";
    private Button authenticateBtn;
    private Executor executor;
    private BiometricPrompt biometricPrompt;
    private BiometricPrompt.PromptInfo promptInfo;
    private TextView authticationStatus;
    private static final int BIOMETRIC_PERMISSION_CODE = 102;
    private String pinHash;
    private String pinSalt;
    private TextView pinInput;
    private TextView pinVerify;
    private TextView pinNewInput;
    private Button pinLogin;
    private Button pinSet;
    public static final String DATA = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz|!£$%&/=@#";
    public Random RANDOM = new Random();
    private TextView pinSetNewBtn;
    private TextView resetAll;

    @RequiresApi(api = Build.VERSION_CODES.P)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_fingerprint);
        authenticateBtn = (Button) findViewById(R.id.authenticate);
        authticationStatus = (TextView) findViewById(R.id.authenticationStatus);
        pinInput = (TextView) findViewById(R.id.pinNumber);
        pinVerify = (TextView) findViewById(R.id.pinNumberVerify);
        pinLogin = (Button) findViewById(R.id.pinAuthenticateBtn);
        pinSet = (Button) findViewById(R.id.pinAuthenticateBtnNew);
        pinSetNewBtn = (TextView) findViewById(R.id.pinsetNewBtn);
        pinNewInput = (TextView) findViewById(R.id.pinNewNumber);
        resetAll = (TextView) findViewById(R.id.pinresetAll);
        if (readPassword()) {
            startActivity(new Intent(FingerprintActivity.this, MainActivity.class));
        }
        ;

        executor = ContextCompat.getMainExecutor(this);
        biometricPrompt = new BiometricPrompt(FingerprintActivity.this, executor, new BiometricPrompt.AuthenticationCallback() {
            @Override
            public void onAuthenticationError(int errorCode, @NonNull CharSequence errString) {
                super.onAuthenticationError(errorCode, errString);
                String text = (String) errString;

            }

            @Override
            public void onAuthenticationSucceeded(@NonNull BiometricPrompt.AuthenticationResult result) {
                super.onAuthenticationSucceeded(result);
                grantAccess();
            }

            @Override
            public void onAuthenticationFailed() {
                super.onAuthenticationFailed();
                String text = "Fehler in Authentifizierung";
                notifyUser(text);
            }
        });
        promptInfo = new BiometricPrompt.PromptInfo.Builder()
                .setTitle(getResources().getString(R.string.fingerprintAuthenticateTitle))
                .setSubtitle(getResources().getString(R.string.fingerprintAuthenticateSubtitle))
                .setNegativeButtonText(getResources().getString(R.string.fingerprintAuthenticateUsePin))
                .build();

        authenticateBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                biometricPrompt.authenticate(promptInfo);
            }
        });
        pinLogin.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (verifyPassword(pinInput.getText().toString())) {
                    grantAccess();
                }
            }
        });
        pinSet.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                savePassword();
            }
        });
        pinSetNewBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                pinVerify.setVisibility(View.VISIBLE);
                pinSet.setVisibility(View.VISIBLE);
                pinSetNewBtn.setVisibility(View.GONE);
                pinLogin.setVisibility(View.GONE);
                pinNewInput.setVisibility(View.VISIBLE);
            }
        });
        resetAll.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                new AlertDialog.Builder(FingerprintActivity.this)
                        .setTitle(R.string.resetTitle)
                        .setMessage(R.string.resetSubtitle)

                        // Specifying a listener allows you to take an action before dismissing the dialog.
                        // The dialog is automatically dismissed when a dialog button is clicked.
                        .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int which) {
                                if (resetAll()) {
                                    Toast.makeText(FingerprintActivity.this, R.string.resetApp, Toast.LENGTH_SHORT).show();

                                }
                            }
                        })

                        // A null listener allows the button to dismiss the dialog and take no further action.
                        .setNegativeButton(android.R.string.no, null)
                        .setIcon(android.R.drawable.ic_dialog_alert)
                        .show();


            }
        });
        if (!pinSalt.equals("")) {
            biometricPrompt.authenticate(promptInfo);
        }

        checkPermission(Manifest.permission.USE_BIOMETRIC,
                BIOMETRIC_PERMISSION_CODE);
    }

    private void grantAccess() {
        String text = getResources().getString(R.string.accessGranted);
        notifyUser(text);
        pinInput.setText("");
        startActivity(new Intent(FingerprintActivity.this, MainActivity.class));
    }

    private void notifyUser(String text) {
        Toast.makeText(this, text, Toast.LENGTH_SHORT).show();
    }

    public void checkPermission(String permission, int requestCode) {
        if (ContextCompat.checkSelfPermission(FingerprintActivity.this, permission)
                == PackageManager.PERMISSION_DENIED) {

            // Requesting the permission
            ActivityCompat.requestPermissions(FingerprintActivity.this,
                    new String[]{permission},
                    requestCode);
        }

    }

    public boolean readPassword() {
        SharedPreferences sharedPreferences = getSharedPreferences(SECURITY_SAVE, MODE_PRIVATE);
        pinHash = sharedPreferences.getString(SaveSettings.USER_PIN_HASH, "");
        pinSalt = sharedPreferences.getString(SaveSettings.USER_PIN_SALT, "");
        if (pinSalt != null && pinSalt.equals("")) {
            pinNewInput.setVisibility(View.VISIBLE);
            pinSet.setVisibility(View.VISIBLE);
            pinVerify.setVisibility(View.VISIBLE);
            pinSetNewBtn.setVisibility(View.GONE);
            pinInput.setVisibility(View.GONE);
            pinLogin.setVisibility(View.GONE);
            authenticateBtn.setVisibility(View.GONE);
        }
        sharedPreferences = getSharedPreferences(MainActivity.SHARED_PREFS, MODE_PRIVATE);
        return sharedPreferences.getBoolean(SaveSettings.IS_ORG, false);
    }

    private boolean verifyPassword(String passwd) {
        String hashAct = digest(pinSalt + passwd, "MD5");
        if (hashAct.equals(pinHash)) {
            return true;
        } else {
            Toast.makeText(this, R.string.pinWrong, Toast.LENGTH_SHORT).show();
        }
        return false;

    }

    public void savePassword() {
        SharedPreferences sharedPreferences = getSharedPreferences(SECURITY_SAVE, MODE_PRIVATE);
        String pin0 = pinInput.getText().toString();
        String pin1 = pinNewInput.getText().toString();
        String pin2 = pinVerify.getText().toString();

        if (!pinSalt.equals("") && !pinHash.equals(digest(pinSalt + pin0, "MD5"))) {
            Toast.makeText(this, R.string.pinWrong, Toast.LENGTH_SHORT).show();
            return;
        }
        if (!pin1.equals(pin2)) {
            Toast.makeText(this, R.string.pinWrong, Toast.LENGTH_SHORT).show();
            return;
        }
        if (pin1.length() < 4) {
            Toast.makeText(this, R.string.pinToShort, Toast.LENGTH_SHORT).show();
            return;
        }
        pinSalt = randomString(24);
        pinHash = digest(pinSalt + (String) pin1, "MD5");
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(SaveSettings.USER_PIN_HASH, pinHash);
        editor.putString(SaveSettings.USER_PIN_SALT, pinSalt);
        editor.apply();
        pinVerify.setVisibility(View.GONE);
        pinNewInput.setVisibility(View.GONE);
        pinSet.setVisibility(View.GONE);
        pinInput.setVisibility(View.VISIBLE);
        pinLogin.setVisibility(View.VISIBLE);
        pinSetNewBtn.setVisibility(View.VISIBLE);
        authenticateBtn.setVisibility(View.VISIBLE);
        Toast.makeText(this, R.string.pinSaved, Toast.LENGTH_SHORT).show();
    }


    public static String digest(String s, String algorithm) {
        MessageDigest m = null;
        try {
            m = MessageDigest.getInstance(algorithm);
        } catch (NoSuchAlgorithmException e) {
            e.printStackTrace();
            return s;
        }

        m.update(s.getBytes(), 0, s.length());
        return byteArrayToHex(m.digest());
    }

    private static char[] hextable = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};

    public static String byteArrayToHex(byte[] array) {
        String s = "";
        for (int i = 0; i < array.length; ++i) {
            int di = (array[i] + 256) & 0xFF; // Make it unsigned
            s = s + hextable[(di >> 4) & 0xF] + hextable[di & 0xF];
        }
        return s;
    }

    public String randomString(int len) {
        StringBuilder sb = new StringBuilder(len);

        for (int i = 0; i < len; i++) {
            sb.append(DATA.charAt(RANDOM.nextInt(DATA.length())));
        }

        return sb.toString();
    }

    public boolean resetAll() {
        SharedPreferences sharedPreferences = getSharedPreferences(SECURITY_SAVE, MODE_PRIVATE);
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.clear();
        editor.apply();
        sharedPreferences = getSharedPreferences(MainActivity.SHARED_PREFS, MODE_PRIVATE);
        editor = sharedPreferences.edit();
        editor.clear();
        editor.apply();
        finish();
        startActivity(getIntent());
        return true;
    }
}

File: SKIBin_old/app/src/main/java/com/h2Invent/skibin/FingerprintActivity.kt
package com.h2Invent.skibin

import android.content.Intent
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.Random

class FingerprintActivity : AppCompatActivity() {
    private lateinit var authenticateButton: Button
    private lateinit var pinInput: EditText
    private lateinit var pinVerify: EditText
    private lateinit var pinNewInput: EditText
    private lateinit var pinLogin: Button
    private lateinit var pinSet: Button
    private lateinit var pinSetNewButton: TextView
    private lateinit var resetAll: TextView

    private var pinHash: String = ""
    private var pinSalt: String = ""

    private lateinit var biometricPrompt: BiometricPrompt
    private lateinit var promptInfo: BiometricPrompt.PromptInfo

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_fingerprint)

        if (!hasSavedConnection()) {
            startActivity(Intent(this, MainActivity::class.java))
            finish()
            return
        }

        authenticateButton = findViewById(R.id.authenticate)
        pinInput = findViewById(R.id.pinNumber)
        pinVerify = findViewById(R.id.pinNumberVerify)
        pinNewInput = findViewById(R.id.pinNewNumber)
        pinLogin = findViewById(R.id.pinAuthenticateBtn)
        pinSet = findViewById(R.id.pinAuthenticateBtnNew)
        pinSetNewButton = findViewById(R.id.pinsetNewBtn)
        resetAll = findViewById(R.id.pinresetAll)

        readPassword()
        setupBiometric()
        updatePinUi(hasStoredPin())

        authenticateButton.setOnClickListener { biometricPrompt.authenticate(promptInfo) }
        pinLogin.setOnClickListener {
            if (verifyPassword(pinInput.text?.toString().orEmpty())) grantAccess()
        }
        pinSet.setOnClickListener { savePassword() }
        pinSetNewButton.setOnClickListener { updatePinUi(false) }
        resetAll.setOnClickListener { confirmResetAll() }

        if (hasStoredPin()) biometricPrompt.authenticate(promptInfo)
    }

    private fun setupBiometric() {
        val executor = ContextCompat.getMainExecutor(this)
        biometricPrompt = BiometricPrompt(this, executor, object : BiometricPrompt.AuthenticationCallback() {
            override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                super.onAuthenticationSucceeded(result)
                grantAccess()
            }

            override fun onAuthenticationFailed() {
                super.onAuthenticationFailed()
                Toast.makeText(this@FingerprintActivity, R.string.pinWrong, Toast.LENGTH_SHORT).show()
            }
        })

        promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(getString(R.string.fingerprintAuthenticateTitle))
            .setSubtitle(getString(R.string.fingerprintAuthenticateSubtitle))
            .setNegativeButtonText(getString(R.string.fingerprintAuthenticateUsePin))
            .build()
    }

    private fun grantAccess() {
        pinInput.setText("")
        Toast.makeText(this, R.string.accessGranted, Toast.LENGTH_SHORT).show()
        startActivity(Intent(this, MainActivity::class.java))
        finish()
    }

    private fun confirmResetAll() {
        AlertDialog.Builder(this)
            .setTitle(R.string.resetTitle)
            .setMessage(R.string.resetSubtitle)
            .setPositiveButton(android.R.string.yes) { _, _ ->
                if (resetEverything()) Toast.makeText(this, R.string.resetApp, Toast.LENGTH_SHORT).show()
            }
            .setNegativeButton(android.R.string.no, null)
            .show()
    }

    private fun hasSavedConnection(): Boolean {
        val appPrefs = getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE)
        return appPrefs.getBoolean(SaveSettings.IS_ORG, false) || appPrefs.getBoolean(SaveSettings.IS_USER, false)
    }

    private fun readPassword() {
        val sharedPreferences = getSharedPreferences(SaveSettings.SECURITY_PREFS, MODE_PRIVATE)
        pinHash = sharedPreferences.getString(SaveSettings.USER_PIN_HASH, "").orEmpty()
        pinSalt = sharedPreferences.getString(SaveSettings.USER_PIN_SALT, "").orEmpty()
    }

    private fun hasStoredPin(): Boolean = pinHash.isNotBlank() && pinSalt.isNotBlank()

    private fun verifyPassword(password: String): Boolean {
        if (digest(pinSalt + password, "MD5") == pinHash) return true
        Toast.makeText(this, R.string.pinWrong, Toast.LENGTH_SHORT).show()
        return false
    }

    private fun savePassword() {
        val currentPin = pinInput.text?.toString().orEmpty()
        val newPin = pinNewInput.text?.toString().orEmpty()
        val verifyPin = pinVerify.text?.toString().orEmpty()

        if (hasStoredPin() && digest(pinSalt + currentPin, "MD5") != pinHash) {
            Toast.makeText(this, R.string.pinWrong, Toast.LENGTH_SHORT).show()
            return
        }
        if (newPin != verifyPin) {
            Toast.makeText(this, R.string.pinWrong, Toast.LENGTH_SHORT).show()
            return
        }
        if (newPin.length < 4) {
            Toast.makeText(this, R.string.pinToShort, Toast.LENGTH_SHORT).show()
            return
        }

        pinSalt = randomString(24)
        pinHash = digest(pinSalt + newPin, "MD5")
        getSharedPreferences(SaveSettings.SECURITY_PREFS, MODE_PRIVATE).edit().apply {
            putString(SaveSettings.USER_PIN_HASH, pinHash)
            putString(SaveSettings.USER_PIN_SALT, pinSalt)
            apply()
        }
        Toast.makeText(this, R.string.pinSaved, Toast.LENGTH_SHORT).show()
        updatePinUi(true)
    }

    private fun updatePinUi(hasPin: Boolean) {
        pinInput.visibility = android.view.View.VISIBLE
        pinLogin.visibility = android.view.View.VISIBLE
        authenticateButton.visibility = if (hasPin) android.view.View.VISIBLE else android.view.View.GONE
        pinNewInput.visibility = if (hasPin) android.view.View.GONE else android.view.View.VISIBLE
        pinVerify.visibility = if (hasPin) android.view.View.GONE else android.view.View.VISIBLE
        pinSet.visibility = if (hasPin) android.view.View.GONE else android.view.View.VISIBLE
        pinSetNewButton.visibility = if (hasPin) android.view.View.VISIBLE else android.view.View.GONE
    }

    private fun resetEverything(): Boolean {
        getSharedPreferences(SaveSettings.SECURITY_PREFS, MODE_PRIVATE).edit().clear().apply()
        getSharedPreferences(SaveSettings.SHARED_PREFS, MODE_PRIVATE).edit().clear().apply()
        startActivity(Intent(this, MainActivity::class.java))
        finish()
        return true
    }

    private fun digest(input: String, algorithm: String): String {
        return try {
            val messageDigest = MessageDigest.getInstance(algorithm)
            messageDigest.update(input.toByteArray(), 0, input.length)
            byteArrayToHex(messageDigest.digest())
        } catch (_: NoSuchAlgorithmException) {
            input
        }
    }

    private fun byteArrayToHex(array: ByteArray): String = buildString {
        array.forEach { append(HEX_TABLE[(it.toInt() shr 4) and 0x0F]).append(HEX_TABLE[it.toInt() and 0x0F]) }
    }

    private fun randomString(length: Int): String = buildString {
        repeat(length) { append(DATA[random.nextInt(DATA.length)]) }
    }

    companion object {
        private const val DATA = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz|!£$%&/=@#"
        private val HEX_TABLE = charArrayOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')
        private val random = Random()
    }
}
